### PR TITLE
gradient detach

### LINKED ISSUE #225: Detach graph for autoregressive outputs
This would allow increased batch sizes

### PR DESCRIPTION
**Pull request by amogh-gulati**
closes #225

### CODE REVIEW

File: src/ocean_emulators/train.py [hunk 1]
```diff
@@ -224,6 +224,7 @@ def __init__(self, cfg: TrainConfig) -> None:
             wet=self.wet.to(self.device),
             area_weights=self.area_weights,
             static_data=self.static_data,
+            gradient_detach_interval=cfg.gradient_detach_interval,
```
Comment (alxmrs) [inline src/ocean_emulators/train.py]: If it's in cfg, I think this var is accessible to the build method (as a class member) and doesn't need to be passed in this way.

File: src/ocean_emulators/config.py [hunk 2]
```diff
@@ -407,6 +407,7 @@ def build(
         wet: Grid,
         area_weights: Grid,
         static_data: xr.Dataset | None,
+        gradient_detach_interval: int = 0,
```
Comment (alxmrs) [inline src/ocean_emulators/config.py]: Why not make it a class member instead of an argument to the build method?

Comment (alxmrs) [reply src/ocean_emulators/config.py]: Further:  I think this could benefit from a `Field` so we can add a docstring.

File: src/ocean_emulators/config.py [hunk 3]
```diff
@@ -427,6 +428,7 @@ def build(
...
```
Comment (alxmrs) [inline src/ocean_emulators/config.py]: If we make it a class member, then I think we can omit them for these args.

File: tests/test_gradient_detaching.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+from ocean_emulators.config import SamudraConfig, UNetBackboneConfig
+from ocean_emulators.constants import TensorMap
+from ocean_emulators.datasets import TrainData
+from ocean_emulators.utils.data import DataSource, Normalize
+from ocean_emulators.utils.multiton import MultitonScope
+
+
+@pytest.fixture
+def samudra_setup():
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: I like this fixture!

File: tests/test_gradient_detaching.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            train_data.insert(input_tensor, label_tensor)
+
+        return create_samudra_model, train_data
+
+
+class TestSamudraGradientDetaching:
+    def test_samudra_forward_pass_no_detaching(self, samudra_setup):
+        """Test Samudra forward pass without gradient detaching."""
+        create_model, train_data = samudra_setup
+        model = create_model(gradient_detach_interval=0)
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: This is a nice interface.

File: tests/test_gradient_detaching.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    grad_count = sum(1 for p in model.parameters() if p.grad is not None)
+    total_params = sum(1 for _ in model.parameters())
+    assert grad_count > 0, "Model should have gradients after backward pass"
+    assert grad_count == total_params, (
+        f"Expected all {total_params} parameters to have gradients, got {grad_count}"
+    )
+
+
+if __name__ == "__main__":
+    _run_standalone_test()
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: I think we should rely on test runners (pytest) instead of making this a runnable script. How much additional value does this standalone test add? Could we just make it a `test_`?

**Review by alxmrs — CHANGES_REQUESTED** [review]
I have an alternative proposal for how to configure this change. But, the core of the change LGTM. Thank you, and I like the tests.

File: src/ocean_emulators/config.py (hunk 3 above)
Comment (amogh-gulati) [reply src/ocean_emulators/config.py]: both samudra and the base model dont have access to the entire config, I think only the required args are being shared

File: src/ocean_emulators/config.py (hunk 2 above)
Comment (amogh-gulati) [reply src/ocean_emulators/config.py]: done!

File: tests/test_gradient_detaching.py (hunk 6 above)
Comment (amogh-gulati) [reply tests/test_gradient_detaching.py]: I was using that as a scratch pad, a test runner makes sense, yes!

File: src/ocean_emulators/config.py [hunk 7]
```diff
@@ -418,6 +418,10 @@ class SamudraConfig(BaseModelConfig):
         default=0,
         description="""Number of channels used for a learned positional embedding""",
     )
+    gradient_detach_interval: int = Field(
+        default=0,
+        description="""Interval for detaching gradients in autoregressive training. 0 means no detaching.""",
```
Comment (alxmrs) [inline src/ocean_emulators/config.py]:
```suggestion
        description="""Interval for detaching gradients in autoregressive training. `0` means no detaching.""",
```

File: src/ocean_emulators/models/base.py [hunk 8]
```diff
@@ -22,6 +22,7 @@ def __init__(
         last_kernel_size,
         pad,
         static_data,
+        gradient_detach_interval: int = 0,
```
Comment (alxmrs) [inline src/ocean_emulators/models/base.py]: Let's not set a default value here. Ideally, we want one source of truth for the config. That it might thrown an error when we forget to set this is a good thing. We used to do this and it caused hard to track down bugs.

File: src/ocean_emulators/models/samudra.py [hunk 9]
```diff
@@ -22,6 +22,7 @@ def __init__(
         hist: int,
         wet: Grid,
         static_data: xr.Dataset | None,
+        gradient_detach_interval: int = 0,
```
Comment (alxmrs) [inline src/ocean_emulators/models/samudra.py]: See note about not setting default vals.

File: tests/test_gradient_detaching.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        h, w = 8, 8
+        coords = {
+            "lev": [0],
+            "lat": (["y"], torch.linspace(-90, 90, h).numpy()),
+            "lon": (["x"], torch.linspace(-180, 180, w).numpy()),
+        }
+        data = xr.Dataset(
+            {
+                "thetao": (["lev", "y", "x"], torch.randn(1, h, w).numpy()),
+                "hfds": (["y", "x"], torch.randn(h, w).numpy()),
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: 🐑 nit: these array creation functions could replace `torch` with `np` to be more succinct.

File: tests/test_gradient_detaching.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            },
+            coords=coords,
+        )
+        ones = xr.Dataset(
+            {
+                "thetao": (["lev", "y", "x"], torch.ones(1, h, w).numpy()),
+                "hfds": (["y", "x"], torch.ones(h, w).numpy()),
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: Same here

File: tests/test_gradient_detaching.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                dilation=[1, 1],
+                n_layers=[1, 1],
+            ),
+            pos_channels=0,
+        )
+
+        def create_samudra_model(gradient_detach_interval=0):
+            # Set the gradient_detach_interval on the config object
+            config.gradient_detach_interval = gradient_detach_interval
+            return config.build(
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: 🐑 instead of changing the gradient detach attr by side effect, I think it's safe to simply move the config instantiation code inside of this function. 
```suggestion

        def create_samudra_model(gradient_detach_interval=0):
            config = SamudraConfig(
                unet=UNetBackboneConfig(
                    ch_width=[4, 8],
                    dilation=[1, 1],
                    n_layers=[1, 1],
                ),
                pos_channels=0,
                gradient_detach_interval=gradient_detach_interval,
            )
            return config.build(
```

File: tests/test_gradient_detaching.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        train_data = TrainData(num_prognostic_channels=1)
+        for step in range(4):
+            input_tensor = torch.randn(1, 2, h, w, requires_grad=True)
+            label_tensor = torch.randn(1, 1, h, w)
+            train_data.insert(input_tensor, label_tensor)
+
+        return create_samudra_model, train_data
+
+
+class TestSamudraGradientDetaching:
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: A lot of these tests are the same, but just have gradient detaching turned on. This is great: it means we can make the gradient detaching options (i.e. `[0, 1]`) into a fixture and then just pass the into the tests. This combinatorial approach should cut the number of test methods in half (and will let you succinctly add more combinations later -- ofc, these could make test time longer.)

**Review by alxmrs — COMMENTED** [review]
A small note about default vals, and then I've added feedback to your test code.

File: tests/test_gradient_detaching.py (hunk 12 above)
Comment (alxmrs) [reply tests/test_gradient_detaching.py]: If you want to be really suave, you could do this, too:

```suggestion

        def create_samudra_model(gradient_detach_interval=0):
            return SamudraConfig(
                unet=UNetBackboneConfig(
                    ch_width=[4, 8],
                    dilation=[1, 1],
                    n_layers=[1, 1],
                ),
                pos_channels=0,
                gradient_detach_interval=gradient_detach_interval,
            ).build(
```

File: src/ocean_emulators/config.py (hunk 7 above)
Comment (amogh-gulati) [reply src/ocean_emulators/config.py]: gotcha

File: src/ocean_emulators/models/base.py (hunk 8 above)
Comment (amogh-gulati) [reply src/ocean_emulators/models/base.py]: yup this makes sense!

File: src/ocean_emulators/config.py [hunk 14]
```diff
@@ -487,6 +487,10 @@ class SamudraConfig(BaseModelConfig):
         default=0,
         description="""Number of channels used for a learned positional embedding""",
     )
+    gradient_detach_interval: int = Field(
+        default=0,
+        description="""Interval for detaching gradients in autoregressive training. `0` means no detaching.""",
+    )
```
Comment (alxmrs) [inline src/ocean_emulators/config.py]: If it's going here and the FOMO model, then I think we could move this up to the BaseModelConfig -- I think it should live next to `checkpointing`, for example. This saves us from having to write the config twice.

File: src/ocean_emulators/models/base.py [hunk 15]
```diff
@@ -22,6 +22,7 @@ def __init__(
         last_kernel_size,
         pad,
         static_data,
+        gradient_detach_interval: int,
```
Comment (alxmrs) [inline src/ocean_emulators/models/base.py:25]: Indeed, since this argument lives in `BaseModel`, it makes even more sense to make the config live in `BaseModelConfig`.

File: tests/test_gradient_detaching.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+from ocean_emulators.config import SamudraConfig, UNetBackboneConfig
+from ocean_emulators.constants import TensorMap
+from ocean_emulators.datasets import TrainData
+from ocean_emulators.utils.data import DataSource, Normalize
+from ocean_emulators.utils.multiton import MultitonScope
+
+
+@pytest.fixture(params=[0, 1, 2])
+def gradient_detach_interval(request):
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py:14]: Are these quick tests? Do they finish fast? If so, then let's definitely test 0, 1, and 2. If they're slow, then I'd be hesitant to include all three cases. In this project, I've been guilty of promoting more test coverage through combinations like this at the cost of having tests take a long time. 

At a high level, this fixture looks great.

File: tests/test_gradient_detaching.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            return model, train_data
+
+    return _create_model_helper
+
+
+@pytest.fixture
+def samudra_setup(gradient_detach_interval, create_samudra_model):
+    """Parametrized fixture that uses the factory to create models."""
+    model, train_data = create_samudra_model(gradient_detach_interval)
+    return model, train_data, gradient_detach_interval
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: I appreciate what you could be trying to do here wrt removing repeated calls, but in my view, this is a bit too DRY. I think it's perfectly OK for each test to accept two fixtures (`create_samudra_model` and `gradient_detatch_interval`) and repeat the function invocation. Looking at the tests, it appears that we need to use too many of the intermediate values. That we accept arguments and immediately return them makes me think that we would be better off omitting this fixture.

File: tests/test_gradient_detaching.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+from ocean_emulators.constants import TensorMap
+from ocean_emulators.datasets import TrainData
+from ocean_emulators.utils.data import DataSource, Normalize
+from ocean_emulators.utils.multiton import MultitonScope
+
+
+@pytest.fixture(params=[0, 1, 2])
+def gradient_detach_interval(request):
+    """Parametrized fixture for gradient detach intervals. `0` means no detaching."""
+    return request.param
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: 🐑 Looking at how this fixture is used, here's an idea: what if this returned tuples of the interval + the "name" of the interval, i.e. if the gradient is detached or not. It could look like this:
```suggestion
    return (i := request.param), "no detaching" if i == 0 else f"detaching every {i}"
```

I am not sure if this is a good fit, but I bet it would clean up some of the logic in your tests code.

File: tests/test_gradient_detaching.py [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+
+@pytest.fixture
+def samudra_setup(gradient_detach_interval, create_samudra_model):
+    """Parametrized fixture that uses the factory to create models."""
+    model, train_data = create_samudra_model(gradient_detach_interval)
+    return model, train_data, gradient_detach_interval
+
+
+class TestSamudraGradientDetaching:
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py]: 🐑  this is totally optional: What do you think about omitting the class? Most of the pytest code I've seen doesn't group things together in classes. Or, when it does, it's because a test file includes lots of tests. Since the tests now just include two functions with combinations, we could keep things really simple and just have the tests be two functions.

Comment (alxmrs) [reply tests/test_gradient_detaching.py]: (totally a matter of taste, and I leave the decision up to you).

File: tests/test_gradient_detaching.py [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+def gradient_detach_interval(request):
+    """Parametrized fixture for gradient detach intervals. `0` means no detaching."""
+    return request.param
+
+
+@pytest.fixture
+def create_samudra_model():
+    """Factory fixture for creating Samudra models with different gradient detach intervals."""
+
+    def _create_model_helper(gradient_detach_interval: int):
```
Comment (alxmrs) [inline tests/test_gradient_detaching.py:23]: I like this use of currying.

**Review by alxmrs — COMMENTED** [review]
A few more notes, but this is looking really sharp. Thanks for the change, and I'm happy we could discuss some conventions in this codebase that aren't really written down anywhere else. Tests are looking good! I love succinct combinatorial testing with pytest.

File: tests/test_gradient_detaching.py (hunk 18 above)
Comment (amogh-gulati) [reply tests/test_gradient_detaching.py]: this was helpful!

File: tests/test_gradient_detaching.py (hunk 16 above)
Comment (amogh-gulati) [reply tests/test_gradient_detaching.py:14]: all 6 of the tests only take 0.51s on my local setup

**Review by alxmrs — APPROVED** [review]
A nice, clean change. Excellent work, and thanks for your patience.
